keel-orca/src/main/kotlin/com/netflix/spinnaker/keel/constraints/CanaryConstraintEvaluator.kt

```diff
@@ -0,0 +1,332 @@
+            orcaService.cancelOrchestration(it.executionId)
+          }
+        }
+      }
+      deliveryConfigRepository.storeConstraintState(
+        state.copy(
+          status = ConstraintStatus.FAIL,
+          judgedBy = judge,
+          judgedAt = clock.instant(),
+          comment = "Canary failed: ${status.summary()}"))
+
+      return false
+    }
+
+    if (status.allPassed() && status.regions() == constraint.regions) {
+      deliveryConfigRepository.storeConstraintState(
+        state.copy(
+          status = ConstraintStatus.PASS,
+          judgedBy = judge,
+          judgedAt = clock.instant(),
+          comment = "Canary passed: ${status.summary()}"))
+
+      return true
+    }
+
+    val regionsToTrigger = shouldTrigger(constraint, attributes)
+    val regionsWithCorrelatedExecutions = runBlocking {
+      constraint.regionsWithCorrelatedExecutions(judge)
+    }
+    val unknownExecutions = regionsWithCorrelatedExecutions.filter {
+      regionsToTrigger.contains(it.key)
+    }
+      .toSortedMap()
+
+    if (unknownExecutions.isNotEmpty() && attributes.executions.isEmpty()) {
+      val message = "$correlatedMessagePrefix ($unknownExecutions), " +
+        "another artifact version may still be under evaluation."
+
+      if (state.comment != message) {
+        deliveryConfigRepository.storeConstraintState(state.copy(comment = message))
+      }
+
+      return false
+    }
+
+    if (regionsToTrigger.isEmpty()) {
+      if (!status.anyRunning()) {
+        deliveryConfigRepository.storeConstraintState(
+          state.copy(
+            status = ConstraintStatus.FAIL,
+            judgedBy = judge,
+            judgedAt = clock.instant(),
+            attributes = attributes.copy(status = status),
+            comment = "Failure encountered launching canaries"))
+      } else if (attributes.status != status) {
+        deliveryConfigRepository.storeConstraintState(
+          state.copy(
+            attributes = attributes.copy(status = status)))
+      }
+
+      return false
+    }
+
+    attributes = attributes.copy(startAttempt = attributes.startAttempt + 1)
+
+    val tasks = deployHandler.deployCanary(
+      constraint = constraint,
+      version = version,
+      deliveryConfig = deliveryConfig,
+      targetEnvironment = targetEnvironment,
+      regions = regionsToTrigger)
+
+    attributes = attributes.copy(
+      executions = tasks.map { (region, task) ->
+        RegionalExecutionId(region, task.id)
+      }
+        .toSet())
+
+    deliveryConfigRepository.storeConstraintState(
+      state.copy(
+        attributes = attributes,
+        comment = "Running canaries in ${attributes.launchedRegions}"))
+
+    return false
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private fun canaryStatus(attributes: CanaryConstraintAttributes?): Set<CanaryStatus> {
+    if (attributes?.executions.isNullOrEmpty()) {
+      return emptySet()
+    }
+
+    return attributes!!.executions
```

Comment:
can't this still be null here? because the if check above only checks if executions is null or empty, not if the overall object is null

Reply:
I think the compiler is not quite grasping the implications of `isNullOrEmpty`. Is that a built in function or one of ours?

Reply:
Emily called it! `isNullOrEmpty` just wasn't applicable to `attributes`. After changing to `if (attributes == null || attributes.executions.isNullOrEmpty())`, the compiler casts `attributes` to the non-nullable type.